File: src/Backend/Core/Sadara.Domain/Entities/Agent.cs
namespace Sadara.Domain.Entities;

// ═══════════════════════════════════════════════════════════════
// 🏪 نظام الوكلاء - Agent System
// الوكيل يدخل بوابة المواطن ويجري عمليات
// تظهر في شاشة الشركة + مدير النظام
// ═══════════════════════════════════════════════════════════════

/// <summary>
/// نوع الوكيل
/// </summary>
public enum AgentType
{
    /// <summary>وكيل خاص</summary>
    Private = 0,
    
    /// <summary>وكيل عام</summary>
    Public = 1
}

/// <summary>
/// حالة الوكيل
/// </summary>
public enum AgentStatus
{
    /// <summary>نشط</summary>
    Active = 0,
    
    /// <summary>معلق</summary>
    Suspended = 1,
    
    /// <summary>محظور</summary>
    Banned = 2,
    
    /// <summary>غير مفعل</summary>
    Inactive = 3
}

/// <summary>
/// الوكيل - يعمل كوسيط بين الشركة والمواطنين
/// يدخل بوابة المواطن ويجري العمليات (تسجيل، تحصيل، صيانة...)
/// كل وكيل له حساب مالي (أجور عليه + تسديد منه = صافي)
/// </summary>
public class Agent : BaseEntity<Guid>
{
    // ============ معلومات أساسية ============
    
    /// <summary>كود الوكيل (يتولد تلقائياً) مثل AGT-0001</summary>
    public string AgentCode { get; set; } = string.Empty;
    
    /// <summary>اسم الوكيل</summary>
    public string Name { get; set; } = string.Empty;
    
    /// <summary>نوع الوكيل</summary>
    public AgentType Type { get; set; } = AgentType.Private;
    
    /// <summary>رقم الهاتف (يُستخدم لتسجيل الدخول)</summary>
    public string PhoneNumber { get; set; } = string.Empty;
    
    /// <summary>كلمة المرور المشفرة</summary>
    public string PasswordHash { get; set; } = string.Empty;
    
    /// <summary>البريد الإلكتروني</summary>
    public string? Email { get; set; }
    
    // ============ الموقع ============
    
    /// <summary>المدينة / المحافظة</summary>
    public string? City { get; set; }
    
    /// <summary>المنطقة / الحي</summary>
    public string? Area { get; set; }
    
    /// <summary>العنوان التفصيلي</summary>
    public string? FullAddress { get; set; }
    
    /// <summary>خط العرض</summary>
    public double? Latitude { get; set; }
    
    /// <summary>خط الطول</summary>
    public double? Longitude { get; set; }
    
    // ============ معرف الصفحة والشركة ============
    
    /// <summary>معرف الصفحة (للربط مع نظام خارجي أو رقم صفحة)</summary>
    public string? PageId { get; set; }
    
    /// <summary>الشركة التي يتبع لها الوكيل</summary>
    public Guid CompanyId { get; set; }
    
    // ============ الحالة ============
    
    /// <summary>حالة الوكيل</summary>
    public AgentStatus Status { get; set; } = AgentStatus.Active;
    
    /// <summary>صورة الوكيل</summary>
    public string? ProfileImageUrl { get; set; }
    
    /// <summary>ملاحظات</summary>
    public string? Notes { get; set; }
    
    /// <summary>آخر تسجيل دخول</summary>
    public DateTime? LastLoginAt { get; set; }
    
    // ============ الحساب المالي (ملخص محسوب) ============
    
    /// <summary>إجمالي الأجور (المبالغ المترتبة على الوكيل)</summary>
    public decimal TotalCharges { get; set; } = 0;
    
    /// <summary>إجمالي التسديدات (المبالغ التي دفعها الوكيل)</summary>
    public decimal TotalPayments { get; set; } = 0;
    
    /// <summary>الصافي = الأجور - التسديدات (موجب = عليه، سالب = له)</summary>
    public decimal NetBalance { get; set; } = 0;
    
    // ============ Navigation Properties ============
    
    /// <summary>الشركة</summary>
    public virtual Company? Company { get; set; }
    
    /// <summary>المعاملات المالية</summary>
    public virtual ICollection<AgentTransaction> Transactions { get; set; } = new List<AgentTransaction>();
}

/// <summary>
/// نوع المعاملة المالية
/// </summary>
public enum TransactionType
{
    /// <summary>أجور - مبلغ مترتب على الوكيل (صرف)</summary>
    Charge = 0,
    
    /// <summary>تسديد - الوكيل يدفع من المبلغ المترتب عليه (قبض)</summary>
    Payment = 1,
    
    /// <summary>خصم - خصم من رصيد الوكيل</summary>
    Discount = 2,
    
    /// <summary>تعديل - تعديل يدوي من المدير</summary>
    Adjustment = 3
}

/// <summary>
/// فئة المعاملة - ما سبب هذه المعاملة؟
/// </summary>
public enum TransactionCategory
{
    /// <summary>تسجيل مشترك جديد</summary>
    NewSubscription = 0,
    
    /// <summary>تجديد اشتراك</summary>
    RenewalSubscription = 1,
    
    /// <summary>صيانة</summary>
    Maintenance = 2,
    
    /// <summary>تحصيل فواتير</summary>
    BillCollection = 3,
    
    /// <summary>تركيب جديد</summary>
    Installation = 4,
    
    /// <summary>نقل خدمة</summary>
    ServiceTransfer = 5,
    
    /// <summary>تسديد نقدي</summary>
    CashPayment = 6,
    
    /// <summary>تحويل بنكي</summary>
    BankTransfer = 7,
    
    /// <summary>أخرى</summary>
    Other = 99
}

/// <summary>
/// المعاملة المالية للوكيل
/// كل عملية يقوم بها الوكيل أو تُسجل عليه تُنشئ معاملة هنا
/// </summary>
public class AgentTransaction : BaseEntity<long>
{
    /// <summary>الوكيل صاحب المعاملة</summary>
    public Guid AgentId { get; set; }
    
    /// <summary>نوع المعاملة (أجور / تسديد / خصم / تعديل)</summary>
    public TransactionType Type { get; set; }
    
    /// <summary>فئة المعاملة (سبب العملية)</summary>
    public TransactionCategory Category { get; set; }
    
    /// <summary>المبلغ (دائماً موجب)</summary>
    public decimal Amount { get; set; }
    
    /// <summary>الرصيد بعد المعاملة</summary>
    public decimal BalanceAfter { get; set; }
    
    /// <summary>وصف المعاملة</summary>
    public string Description { get; set; } = string.Empty;
    
    /// <summary>رقم المرجع (مثل رقم الفاتورة أو رقم طلب الخدمة)</summary>
    public string? ReferenceNumber { get; set; }
    
    /// <summary>معرف طلب الخدمة المرتبط (إن وجد)</summary>
    public Guid? ServiceRequestId { get; set; }
    
    /// <summary>معرف المواطن المرتبط (إن وجد)</summary>
    public Guid? CitizenId { get; set; }
    
    /// <summary>من أنشأ المعاملة (مدير / وكيل / نظام)</summary>
    public Guid? CreatedById { get; set; }
    
    /// <summary>ملاحظات إضافية</summary>
    public string? Notes { get; set; }
    
    // ============ Navigation Properties ============
    
    /// <summary>الوكيل</summary>
    public virtual Agent? Agent { get; set; }
    
    /// <summary>طلب الخدمة</summary>
    public virtual ServiceRequest? ServiceRequest { get; set; }
    
    /// <summary>المواطن</summary>
    public virtual Citizen? Citizen { get; set; }
}

// ═══════════════════════════════════════════════════════════════
// 💰 نسب عمولات الوكلاء - Agent Commission Rates
// لكل وكيل × لكل باقة = نسبة مئوية من ربح الباقة
// ═══════════════════════════════════════════════════════════════

/// <summary>
/// نسبة عمولة الوكيل لباقة معينة
/// العمولة = ProfitAmount (من InternetPlan) × CommissionPercentage / 100
/// </summary>
public class AgentCommissionRate : BaseEntity<long>
{
    /// <summary>معرف الوكيل</summary>
    public Guid AgentId { get; set; }
    
    /// <summary>معرف باقة الإنترنت</summary>
    public Guid InternetPlanId { get; set; }
    
    /// <summary>نسبة العمولة المئوية (مثال: 10 = 10%)</summary>
    public decimal CommissionPercentage { get; set; }
    
    /// <summary>معرف الشركة</summary>
    public Guid CompanyId { get; set; }
    
    /// <summary>هل مفعّل؟</summary>
    public bool IsActive { get; set; } = true;
    
    /// <summary>ملاحظات</summary>
    public string? Notes { get; set; }
    
    // ============ Navigation Properties ============
    
    /// <summary>الوكيل</summary>
    public virtual Agent? Agent { get; set; }
    
    /// <summary>باقة الإنترنت</summary>
    public virtual InternetPlan? InternetPlan { get; set; }
    
    /// <summary>الشركة</summary>
    public virtual Company? Company { get; set; }
}

File: src/Apps/CompanyDesktop/alsadara-ftth/lib/models/agent.dart
/// نموذج الوكيل
/// يحتوي على جميع بيانات الوكيل ومحاسبته
library;

/// نوع الوكيل
enum AgentType {
  privateAgent, // خاص
  publicAgent; // عام

  static AgentType fromInt(int value) {
    switch (value) {
      case 0:
        return privateAgent;
      case 1:
        return publicAgent;
      default:
        return privateAgent;
    }
  }

  int get intValue {
    switch (this) {
      case privateAgent:
        return 0;
      case publicAgent:
        return 1;
    }
  }

  String get displayName {
    switch (this) {
      case privateAgent:
        return 'وكيل خاص';
      case publicAgent:
        return 'وكيل عام';
    }
  }
}

/// حالة الوكيل
enum AgentStatus {
  active, // نشط
  suspended, // معلق
  banned, // محظور
  inactive; // غير مفعل

  static AgentStatus fromInt(int value) {
    switch (value) {
      case 0:
        return active;
      case 1:
        return suspended;
      case 2:
        return banned;
      case 3:
        return inactive;
      default:
        return active;
    }
  }

  static AgentStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'active':
        return active;
      case 'suspended':
        return suspended;
      case 'banned':
        return banned;
      case 'inactive':
        return inactive;
      default:
        return active;
    }
  }

  int get intValue {
    switch (this) {
      case active:
        return 0;
      case suspended:
        return 1;
      case banned:
        return 2;
      case inactive:
        return 3;
    }
  }

  String get displayName {
    switch (this) {
      case active:
        return 'نشط';
      case suspended:
        return 'معلق';
      case banned:
        return 'محظور';
      case inactive:
        return 'غير مفعل';
    }
  }
}

/// نوع المعاملة المالية
enum TransactionType {
  charge, // أجور
  payment, // تسديد
  discount, // خصم
  adjustment; // تعديل

  static TransactionType fromInt(int value) {
    switch (value) {
      case 0:
        return charge;
      case 1:
        return payment;
      case 2:
        return discount;
      case 3:
        return adjustment;
      default:
        return charge;
    }
  }

  String get displayName {
    switch (this) {
      case charge:
        return 'أجور';
      case payment:
        return 'تسديد';
      case discount:
        return 'خصم';
      case adjustment:
        return 'تعديل';
    }
  }
}

/// فئة المعاملة
enum TransactionCategory {
  newSubscription,
  renewalSubscription,
  maintenance,
  billCollection,
  installation,
  serviceTransfer,
  cashPayment,
  bankTransfer,
  other;

  static TransactionCategory fromInt(int value) {
    switch (value) {
      case 0:
        return newSubscription;
      case 1:
        return renewalSubscription;
      case 2:
        return maintenance;
      case 3:
        return billCollection;
      case 4:
        return installation;
      case 5:
        return serviceTransfer;
      case 6:
        return cashPayment;
      case 7:
        return bankTransfer;
      default:
        return other;
    }
  }

  int get intValue {
    switch (this) {
      case newSubscription:
        return 0;
      case renewalSubscription:
        return 1;
      case maintenance:
        return 2;
      case billCollection:
        return 3;
      case installation:
        return 4;
      case serviceTransfer:
        return 5;
      case cashPayment:
        return 6;
      case bankTransfer:
        return 7;
      case other:
        return 99;
    }
  }

  String get displayName {
    switch (this) {
      case newSubscription:
        return 'تسجيل مشترك جديد';
      case renewalSubscription:
        return 'تجديد اشتراك';
      case maintenance:
        return 'صيانة';
      case billCollection:
        return 'تحصيل فواتير';
      case installation:
        return 'تركيب جديد';
      case serviceTransfer:
        return 'نقل خدمة';
      case cashPayment:
        return 'تسديد نقدي';
      case bankTransfer:
        return 'تحويل بنكي';
      case other:
        return 'أخرى';
    }
  }
}

/// ============================
/// نموذج الوكيل
/// ============================
class AgentModel {
  final String id;
  final String agentCode;
  final String name;
  final AgentType type;
  final String phoneNumber;
  final String? email;
  final String? city;
  final String? area;
  final String? fullAddress;
  final double? latitude;
  final double? longitude;
  final String? pageId;
  final String companyId;
  final String? companyName;
  final AgentStatus status;
  final String? profileImageUrl;
  final String? notes;
  final String? plainPassword;
  final DateTime? lastLoginAt;
  final double totalCharges;
  final double totalPayments;
  final double netBalance;
  final DateTime createdAt;
  final DateTime? updatedAt;

  AgentModel({
    required this.id,
    required this.agentCode,
    required this.name,
    required this.type,
    required this.phoneNumber,
    this.email,
    this.city,
    this.area,
    this.fullAddress,
    this.latitude,
    this.longitude,
    this.pageId,
    required this.companyId,
    this.companyName,
    required this.status,
    this.profileImageUrl,
    this.notes,
    this.plainPassword,
    this.lastLoginAt,
    this.totalCharges = 0,
    this.totalPayments = 0,
    this.netBalance = 0,
    required this.createdAt,
    this.updatedAt,
  });

  factory AgentModel.fromJson(Map<String, dynamic> json) {
    return AgentModel(
      id: json['id'] ?? '',
      agentCode: json['agentCode'] ?? '',
      name: json['name'] ?? '',
      type: json['typeValue'] != null
          ? AgentType.fromInt(json['typeValue'])
          : AgentType.privateAgent,
      phoneNumber: json['phoneNumber'] ?? '',
      email: json['email'],
      city: json['city'],
      area: json['area'],
      fullAddress: json['fullAddress'],
      latitude: json['latitude']?.toDouble(),
      longitude: json['longitude']?.toDouble(),
      pageId: json['pageId'],
      companyId: json['companyId'] ?? '',
      companyName: json['companyName'],
      status: json['statusValue'] != null
          ? AgentStatus.fromInt(json['statusValue'])
          : AgentStatus.fromString(json['status'] ?? 'active'),
      profileImageUrl: json['profileImageUrl'],
      notes: json['notes'],
      plainPassword: json['plainPassword'],
      lastLoginAt: json['lastLoginAt'] != null
          ? DateTime.tryParse(json['lastLoginAt'])
          : null,
      totalCharges: (json['totalCharges'] ?? 0).toDouble(),
      totalPayments: (json['totalPayments'] ?? 0).toDouble(),
      netBalance: (json['netBalance'] ?? 0).toDouble(),
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'agentCode': agentCode,
      'name': name,
      'type': type.intValue,
      'phoneNumber': phoneNumber,
      'email': email,
      'city': city,
      'area': area,
      'fullAddress': fullAddress,
      'latitude': latitude,
      'longitude': longitude,
      'pageId': pageId,
      'companyId': companyId,
      'status': status.intValue,
      'notes': notes,
      'plainPassword': plainPassword,
      'profileImageUrl': profileImageUrl,
    };
  }

  /// هل عليه ديون؟
  bool get hasDebt => netBalance > 0;

  /// هل له رصيد إيجابي (دائن)؟
  bool get hasCredit => netBalance < 0;

  /// الرصيد المعروض (بالقيمة المطلقة)
  double get displayBalance => netBalance.abs();

  AgentModel copyWith({
    String? name,
    AgentType? type,
    String? phoneNumber,
    String? email,
    String? city,
    String? area,
    String? fullAddress,
    double? latitude,
    double? longitude,
    String? pageId,
    AgentStatus? status,
    String? notes,
    String? profileImageUrl,
  }) {
    return AgentModel(
      id: id,
      agentCode: agentCode,
      name: name ?? this.name,
      type: type ?? this.type,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      email: email ?? this.email,
      city: city ?? this.city,
      area: area ?? this.area,
      fullAddress: fullAddress ?? this.fullAddress,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      pageId: pageId ?? this.pageId,
      companyId: companyId,
      companyName: companyName,
      status: status ?? this.status,
      profileImageUrl: profileImageUrl ?? this.profileImageUrl,
      notes: notes ?? this.notes,
      plainPassword: plainPassword,
      lastLoginAt: lastLoginAt,
      totalCharges: totalCharges,
      totalPayments: totalPayments,
      netBalance: netBalance,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}

/// ============================
/// نموذج المعاملة المالية
/// ============================
class AgentTransactionModel {
  final int id;
  final String agentId;
  final TransactionType type;
  final TransactionCategory category;
  final double amount;
  final double balanceAfter;
  final String description;
  final String? referenceNumber;
  final String? serviceRequestId;
  final String? citizenId;
  final String? createdById;
  final String? notes;
  final String? journalEntryId;
  final String? journalEntryNumber;
  final DateTime createdAt;

  AgentTransactionModel({
    required this.id,
    required this.agentId,
    required this.type,
    required this.category,
    required this.amount,
    required this.balanceAfter,
    required this.description,
    this.referenceNumber,
    this.serviceRequestId,
    this.citizenId,
    this.createdById,
    this.notes,
    this.journalEntryId,
    this.journalEntryNumber,
    required this.createdAt,
  });

  factory AgentTransactionModel.fromJson(Map<String, dynamic> json) {
    return AgentTransactionModel(
      id: json['id'] ?? 0,
      agentId: json['agentId'] ?? '',
      type: TransactionType.fromInt(json['typeValue'] ?? 0),
      category: TransactionCategory.fromInt(json['categoryValue'] ?? 99),
      amount: (json['amount'] ?? 0).toDouble(),
      balanceAfter: (json['balanceAfter'] ?? 0).toDouble(),
      description: json['description'] ?? '',
      referenceNumber: json['referenceNumber'],
      serviceRequestId: json['serviceRequestId'],
      citizenId: json['citizenId'],
      createdById: json['createdById'],
      notes: json['notes'],
      journalEntryId: json['journalEntryId']?.toString(),
      journalEntryNumber: json['journalEntryNumber']?.toString(),
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  /// هل هي عملية خصم من الوكيل (أجور)؟
  bool get isCharge => type == TransactionType.charge;

  /// هل هي عملية تسديد؟
  bool get isPayment => type == TransactionType.payment;
}

/// ============================
/// نموذج ملخص المحاسبة
/// ============================
class AgentAccountingSummary {
  final int totalAgents;
  final int activeAgents;
  final double totalCharges;
  final double totalPayments;
  final double totalNetBalance;
  final int agentsWithDebt;
  final int agentsWithCredit;

  AgentAccountingSummary({
    required this.totalAgents,
    required this.activeAgents,
    required this.totalCharges,
    required this.totalPayments,
    required this.totalNetBalance,
    required this.agentsWithDebt,
    required this.agentsWithCredit,
  });

  factory AgentAccountingSummary.fromJson(Map<String, dynamic> json) {
    return AgentAccountingSummary(
      totalAgents: json['totalAgents'] ?? 0,
      activeAgents: json['activeAgents'] ?? 0,
      totalCharges: (json['totalCharges'] ?? 0).toDouble(),
      totalPayments: (json['totalPayments'] ?? 0).toDouble(),
      totalNetBalance: (json['totalNetBalance'] ?? 0).toDouble(),
      agentsWithDebt: json['agentsWithDebt'] ?? 0,
      agentsWithCredit: json['agentsWithCredit'] ?? 0,
    );
  }
}
